designs/2025-css-vars-tracking/README.md

```diff
@@ -0,0 +1,128 @@
+
+### `getVariableValues()` Method
+
+```ts
+interface CSSSourceCode {
+    getVariableValues(func: FunctionNode): Array<Raw>;
+}
+```
+
+A new public method, `getVariableValues(func)`, will be added to `CSSSourceCode`. This method will take a `var()` `Function` node as an argument and return an array of `Raw` nodes representing the declared values of the custom property. The fallback value, if specified in the `FunctionNode`, is returned as the last element of the array. 
+
+### Initialization
+
+The `traverse()` method in `CSSSourceCode` will be updated to populate the `#customProperties` map and the internal data structure (`WeakMap<DeclarationPlain, Array<FunctionNode>>`) used by `getDeclarationVariables()`. During traversal, it will identify `Declaration` nodes that define custom properties and `Function` nodes that are `var()` calls.
+
+## Documentation
+
+Because we don't provide documentation for `CSSSourceCode`, we will rely primarily on TypeScript types to inform rule developers as to these new class members.
+
+## Drawbacks
+
+The primary drawback of this approach is that it only tracks custom properties within a single file. It cannot resolve custom properties that are defined in other files or via mechanisms like inline styles on HTML elements. For the initial implementation, this is considered an acceptable tradeoff.
+
+## Backwards Compatibility Analysis
+
+This is a new feature and does not change any existing APIs. It is fully backwards compatible.
+
+## Alternatives
+
+The primary alternative is to have each rule implement its own custom property tracking logic. This is the current state of affairs and is what this RFC aims to improve upon. A centralized approach is more efficient and less error-prone.
+
+## Open Questions
+
+**Should we use "var" instead of "variable" in function names?**
+
+The proposed API includes methods like `getDeclarationVariables()` and `getVariableValue()`. Since CSS custom properties are commonly referred to as "CSS variables" in the community, there's a question about whether the method names should use the shorter "var" terminology instead. For example:
+
+- `getDeclarationVars()` instead of `getDeclarationVariables()`
+- `getVarValue()` instead of `getVariableValue()`
+
+Using "var" would be more concise and align with the CSS `var()` function syntax that developers are already familiar with. However, "variable" is more explicit and follows common naming conventions in programming APIs.
+
+**Should `getVariableValues()` return the fallback value?**
+
+The fallback value is already available in the `FunctionNode` passed into `getVariableValues()`, so rule authors can still get access to that value easily. It seems like a nice touch to always have it as the last element of the returned array, but that also means that a non-declared value is present in the array, which could potentially be confusing.
```

Comment:
I'd say no, because that would be confusing. An alternative is returning the fallback value as an additional property, but that seems unnecessary if the fallback can be always accessed from the argument.

Reply:
Okay, leaving it so the fallback value is always the last value.